### PR TITLE
#166840878 User signup error swag

### PR DESCRIPTION
**Pull request by tolumide-ng**
#### What does this PR do?
- Enables user to signup on errorswag

#### Description of Task to be completed?
- User can signup on errorswag

#### How should this be manually tested?
- Navigate to the signup page on errorswag

#### What are the relevant pivotal tracker stories?(if applicable)
#166840878

#### Screenshots
[166840878](https://www.pivotaltracker.com/story/show/166840878)

### CODE REVIEW

File: src/pages/Signup/index.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+const SignupForm = ({ isLoading, error, isCompleted, signUp }) => (
+  // whole form
+  <div className="w-7/12 mx-auto m-16 rounded-lg flex h-full">
+    {/* olorunwa's part simulated */}
+    <div className="flex h-100 w-6/12">Welcome here</div>
+    {/* form part */}
+    <div className="w-7/12 p-10">
+      {/* logo and inputs */}
+      <div className="w-full flex flex-col items-center p-2 mx-auto">
+        {/* <div className="w-5/6 mx-auto bg-white mt-10 mb-4 text-center flex flex-col items-center"> */}
```
Comment (Onyimatics) [inline src/pages/Signup/index.js]: you may need to remove these comments please.

File: src/fakeData.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      read: true,
+      payload: {
+        username: 'Adebola',
+        title: 'Are you still coming?',
+        slug: 'theArticleslug',
+        type: 'new_article'
+      }
+    }
+  ]
+};
```
Comment (Onyimatics) [inline src/fakeData.js]: Please what is the value of this file to this PR?

**Review by Onyimatics — CHANGES_REQUESTED** [review]
Good job @tolumide-ng and @olorunwalawrence.
Please endeavour to remove all the unneeded comments on your code.

File: tests/signup.spec.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  signupReducer: {
+    isLoading: false,
+    error: {},
+    isCompleted: false,
+    signUp: jest.fn()
+  }
+});
+const fake = jest.fn();
+
+describe('it should work', () => {
```
Comment (segunolalive) [inline tests/signup.spec.js]: This should be the component being described

File: tests/signup.spec.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  });
+
+  it('If it works', () => {
+    expect(toJson(wrapper)).toMatchSnapshot();
+  });
+
+  it('renders the component', () => {
+    expect(wrapper.exists()).toBe(true);
+    expect(wrapper.find('Form').exists()).toBe(true);
+    expect(wrapper.find('button').exists()).toBe(true);
```
Comment (segunolalive) [inline tests/signup.spec.js]: I think asserting that there's a submit button (type=submit)/ the text in the button is more useful than just that there's a button.

Assert the number of buttons. Other important form elements as well.

Comment (segunolalive) [reply tests/signup.spec.js]: This test doesn't actually test that the component signs a user up. And that is reasonable because the component doesn't sign a user up. The action/network request does that. The component only displays what is appropriate based on the application state.

Also, add test for negative paths e.g when a user enters invalid data.

**Hypothetical Solution**
```js
const  jest.fn(() => 'lol')

state =  { signUp }

// then in your test.

sbmitBtn.simulate('click');

// then you should be able to do:

expect(signUp).toHaveBeenCalled();
expect(signUp).toHaveBeenCalledTimes(1);

```

**Review by segunolalive — CHANGES_REQUESTED** [review]
@olorunwalawrence @tolumide-ng 

This is good. Tests need a bit more work but good job guys.

File: src/pages/Signup/index.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
+import './index.css';
+import React from 'react';
+import { connect } from 'react-redux';
+import { Formik, Form, Field } from 'formik';
+import { signupAction } from '../../store/modules/auth/actions/signup';
+import logo from '../../assets/images/logo.png';
+// import signupCard from '../../assets/images/signup.png';
```
Comment (codexempire) [inline src/pages/Signup/index.js]: Remove comments from the code

File: src/pages/Signup/index.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+import SignupSchema from './schema';
+
+export const SignupForm = ({ isLoading, error, status, signUp }) => (
```
Comment (codexempire) [inline src/pages/Signup/index.js]: Notice you exported the function regularly and also exported the same function as a default would love to know what it's for, but I do think it won't be necessary to export a function regularly and also as the default

File: src/components/App.js [hunk 7]
```diff
@@ -1,12 +1,19 @@
...
+import store from '../store';
 
 const App = () => {
+  useSetUser({ ...store });
   return (
-    <Router>
-      <Routes />
-    </Router>
+    <Provider store={store}>
+      <Router>
```
Comment (codexempire) [inline src/components/App.js:11]: Can you move the `<Router></Router>` wrapping the `<Routes />` into the route folder so as to group similar things together

**Review by codexempire — COMMENTED** [review]
Nice work Noticed the integration of various packages including **`toastr`** left some comment behind

Apart from the comment I also think you should try removing **`.DS_STORE`** from the **`PR`**

**Review by timi-codes — CHANGES_REQUESTED** [review]
Good job @olorunwalawrence @tolumide-ng, However, I noticed your actions and thunk are not properly tested which caused the reduction in your test coverage.
You should ensure the thunks are tested to dispatch the right actions. Also, use the mock store dependency to mock the store and moxios to mock the API requests in your thunk.

You may also want to check out this test example - https://github.com/segunolalive/helloBooks/blob/development/client/__tests__/actions/authActions.spec.js
and this
https://github.com/andela/ah-kifaru-frontend/pull/27/files#diff-8

File: src/store/modules/auth/reducer/index.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+  switch (type) {
+    case AUTH_PENDING:
+      return { ...state, ...payload };
+    case AUTH_SUCCESS:
+      return { ...state, ...payload };
+    case AUTH_FAILURE:
+      return { ...state, ...payload };
+    default:
+      return state;
+  }
```
Comment (segunolalive) [inline src/store/modules/auth/reducer/index.js]: You could have done this since you've done the heavy lifting in your action creators.
```js
function authReducer(state = auth, { type, payload }) {
  return authActions.includes(type) ? { ...state, ...payload } : state;
}

```

File: src/store/modules/auth/actions/index.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  dispatch(authPending());
+  try {
+    const response = await axios({
+      method: 'post',
+      url: `${config.apiUrl}/auth/signup`,
+      data: { username, email, password }
+    });
+    const { token } = response.data.data;
+    saveToLocalStorage({ token, url });
+    const user = decodeToken({ history });
```
Comment (codexempire) [inline src/store/modules/auth/actions/index.js:60]: The after signup the user should be told to verify their accounts and the page may not have to be directed to another. Rather it might just display a message showing successful registration so the user could verify his/her account

**Review by codexempire — CHANGES_REQUESTED** [review]
Nice work please attend to the comment

File: src/store/modules/auth/actions/index.js (hunk 9 above)
Comment (tolumide-ng) [reply src/store/modules/auth/actions/index.js:60]: As much as I agree that a user should be notified to check their email address for the verification link (which can be accomplished via Toastr). I disagree with not redirecting them to the page they were initially on before signing up or the homepage of the application incase they were previously not on an errorSwag page that needed user authentication. Not only does this improve the user experience, it also makes the application more accessible

More so, I think we already discussed this during our state management meeting for authentication last week.

File: src/components/App.js (hunk 7 above)
Comment (tolumide-ng) [reply src/components/App.js:11]: This would not be moved, for the following reasons. 
We need a component called getUser which is already been built by @kechyy at this point to restore the user when a refresh occurs. Also we need to separate concerns across the application to improve a single functionality.

File: src/pages/Signup/index.js (hunk 6 above)
Comment (tolumide-ng) [reply src/pages/Signup/index.js]: Thank you for asking. Please check the test module here for clarification. 
https://github.com/andela/ah-kifaru-frontend/blob/ft-user-signup-errorSwag-166840878/src/pages/Signup/spec.js

**Comment by tolumide-ng** [comment]
> Good job @olorunwalawrence @tolumide-ng, However, I noticed your actions and thunk are not properly tested which caused the reduction in your test coverage.
> You should ensure the thunks are tested to dispatch the right actions. Also, use the mock store dependency to mock the store and moxios to mock the API requests in your thunk.
> 
> You may also want to check out this test example - https://github.com/segunolalive/helloBooks/blob/development/client/__tests__/actions/authActions.spec.js
> and this
> https://github.com/andela/ah-kifaru-frontend/pull/27/files#diff-8

Thank you for your observation. The tests are already been catered for by @kechyy since we are using the same actions and reducers

File: src/components/SocialMedia/social.spec.js [hunk 10]
```diff
@@ -77,7 +79,7 @@ describe('SocialMediaLogin', () => {
       </Provider>
     );
 
-    expect(localStorage.getItem('token')).toEqual(userToken);
+    // expect(localStorage.getItem('token')).toEqual(userToken);
```
Comment (Onyimatics) [inline src/components/SocialMedia/social.spec.js]: I suggest you reach out next time, instead of commenting this `assertion` out.
You may need to replace that assertion with 
```js
expect(JSON.parse(localStorage.getItem('token'))).toEqual(userToken);
Thank you.
```

File: src/pages/Signup/index.spec.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+import { SignupForm } from './index';
+
+const history = { push: jest.fn() };
+const location = { url: '' };
+const defaultProps = {
+  history,
+  location,
+  status: 'authenticationLoading',
+  SignUp: jest.fn(({ userData = {}, history, location }, cb) => {
+    cb({ data: true });
```
Comment (Onyimatics) [inline src/pages/Signup/index.spec.js]: I suggest you consider making the name of this `cb` more descriptive.

File: src/pages/Signup/index.spec.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+  return (
+    <Provider store={args}>
+      <Router>
+        <SignupForm {...defaultProps} />
+      </Router>
+    </Provider>
+  );
+};
+
+describe.only('SignupForm User Input Validation', () => {
```
Comment (Onyimatics) [inline src/pages/Signup/index.spec.js]: I guess you must have forgotten to remove this `.only`

File: src/store/modules/auth/actions/index.js [hunk 13]
```diff
@@ -33,24 +34,35 @@ export const authFailure = error => ({
...
+    Toastr.success('Welcome to ErrorSwag!');
     dispatch(authSuccess(user));
 
-    return url ? history.push(url) : history.push();
-  } catch (error) {
-    const { message } = error.response.data;
+    return url ? history.push(url) : history.push('/articles');
+  } catch ({ response }) {
+    const { message } = response.data;
```
Comment (Onyimatics) [inline src/store/modules/auth/actions/index.js]: instead of removing these lines 
```js
    } catch (error) {
    const { message } = error.response.data;
```, I suggest you use an `if` statement there to also cater for the error coming from `error.response.data` , incase a user encounters a `network error`.

**Review by Onyimatics — CHANGES_REQUESTED** [review]
Well done guys. You may need to check out the comments I left.

File: __test__/App.spec.js [hunk 14]
```diff
@@ -1,6 +1,7 @@
 import React from 'react';
 import { BrowserRouter, Switch } from 'react-router-dom';
 import thunk from 'redux-thunk';
+import { mount } from 'enzyme';
```
Comment (tohbay) [inline __test__/App.spec.js]: Hi @tolumide-ng and @olorunwalawrence,

Please remove the `mount` function since it has been configured globally on the project.

File: __test__/components/navbar.spec.js [hunk 15]
```diff
@@ -1,6 +1,7 @@
 import React from 'react';
 import { Provider } from 'react-redux';
 import configureStore from 'redux-mock-store';
+import { mount } from 'enzyme';
```
Comment (tohbay) [inline __test__/components/navbar.spec.js]: Hi @tolumide-ng and @olorunwalawrence,

Please remove the `mount` function since it has been configured globally on the project.

Comment (olorunwalawrence) [reply __test__/components/navbar.spec.js]: sorry this mount is on navbar component not on the signup

File: enzyme.config.js [hunk 16]
```diff
@@ -0,0 +1,4 @@
+import { configure } from 'enzyme';
```
Comment (tohbay) [inline enzyme.config.js]: I think there is no need for this file `enzyme.config.js`, the configurations therein are already implemented on `__test__/setup.js`

**Review by tohbay — CHANGES_REQUESTED** [review]
Good job @tolumide-ng  and @olorunwalawrence 👍 

There are 4 commits, please squash them. Also, attend to the comments.

File: __test__/App.spec.js (hunk 14 above)
Comment (tolumide-ng) [reply __test__/App.spec.js]: Okay

File: enzyme.config.js (hunk 16 above)
Comment (tolumide-ng) [reply enzyme.config.js]: Okay

**Comment by tolumide-ng** [comment]
> Good job @tolumide-ng and @olorunwalawrence 👍
> 
> There are 4 commits, please squash them. Also, attend to the comments.

Thank you for your review. However, commits cannot be squashed yet as instructed by the TTL since this is a collaborated PR

File: src/components/SocialMedia/social.spec.js (hunk 10 above)
Comment (tolumide-ng) [reply src/components/SocialMedia/social.spec.js]: > I suggest you reach out next time, instead of commenting this `assertion` out.
> You may need to replace that assertion with
> 
> ```js
> expect(JSON.parse(localStorage.getItem('token'))).toEqual(userToken);
> Thank you.
> ```

Hello @Onyimatics , thank you for the review. However, there is no problem with the previous assertion. I only forgot to remove the comments after testing what I needed on this part of the code. Thank you.

File: src/pages/Signup/index.spec.js (hunk 11 above)
Comment (tolumide-ng) [reply src/pages/Signup/index.spec.js]: Okay

File: src/pages/Signup/index.spec.js (hunk 12 above)
Comment (tolumide-ng) [reply src/pages/Signup/index.spec.js]: Yh, I did. Thank you

File: src/store/modules/auth/actions/index.js (hunk 13 above)
Comment (tolumide-ng) [reply src/store/modules/auth/actions/index.js]: Oh True. However, In the case of a network failure, `axios` still returns a response since that it is what was stated on the try block. 
Hence, I think 
``` const message = response && response.data && response.data.message ```
might be okay. What do you think?

File: package.json [hunk 17]
```diff
@@ -92,38 +93,47 @@
...
     "autoprefixer": "^9.6.1",
     "axios": "^0.19.0",
     "clean-webpack-plugin": "^3.0.0",
     "css-loader": "^3.1.0",
     "dotenv-webpack": "^1.7.0",
     "express": "^4.17.1",
+    "formik": "^1.5.8",
     "glob": "^7.1.4",
     "html-webpack-plugin": "^3.2.0",
+    "i": "^0.3.6",
```
Comment (timi-codes) [inline package.json]: kindly remove this as it must have been installed in err.

File: package.json [hunk 18]
```diff
@@ -92,38 +93,47 @@
...
     "jest-svg-transformer": "^1.0.0",
     "jwt-decode": "^2.2.0",
     "mini-css-extract-plugin": "^0.8.0",
+    "npm": "^6.11.0",
```
Comment (timi-codes) [inline package.json]: kindly remove this as it must have been installed in err.

**Comment by tolumide-ng** [comment]
> Good job @olorunwalawrence @tolumide-ng, However, I noticed your actions and thunk are not properly tested which caused the reduction in your test coverage.
> You should ensure the thunks are tested to dispatch the right actions. Also, use the mock store dependency to mock the store and moxios to mock the API requests in your thunk.
> 
> You may also want to check out this test example - https://github.com/segunolalive/helloBooks/blob/development/client/__tests__/actions/authActions.spec.js
> and this
> https://github.com/andela/ah-kifaru-frontend/pull/27/files#diff-8

Thank you @timi-codes . The actions, thunk and reducers are not tested at all in this PR since this feature was implemented in conjunction with the login feature (assigned to @tobechukwuobitube and @kechyy ), they are already testing the actions and reducers appropriately. Thank you once again

File: src/components/App.js (hunk 7 above)
Comment (timi-codes) [reply src/components/App.js:11]: As @codexempire mentioned, the TTL suggested these changes since the App and Router are simple components. Kindly revert back to what we have as @kechyy also have this changes.

File: src/index.css [hunk 19]
```diff
@@ -0,0 +1,9 @@
+@tailwind base;
```
Comment (timi-codes) [inline src/index.css]: Please use the `main.scss`for global stylings

**Review by timi-codes — CHANGES_REQUESTED** [review]
@olorunwalawrence @tolumide-ng Good job on this pull request. However, Kindly attend to all the changes as soon as possible so that we can have this merged today.

**Comment by tolumide-ng** [comment]
> @olorunwalawrence @tolumide-ng Good job on this pull request. However, Kindly attend to all the changes as soon as possible so that we can have this merged today.

Changes attended. Please rereview

File: src/index.css (hunk 19 above)
Comment (tolumide-ng) [reply src/index.css]: I guess, this was when you changed the file configuration to sass as it was initially index.css when I setup the tailwindcss
I would remove that now

File: package.json (hunk 18 above)
Comment (tolumide-ng) [reply package.json]: 👍

File: src/pages/Signup/index.js [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+import { SignupFormComponent } from '@components/SignupForm';
+import { authAction } from '../../store/modules/auth/actions';
+import logo from '../../assets/images/logo.png';
+import SocialMediaLogin from '../../components/SocialMedia/SocialMediaLogin';
+
+export const SignupForm = ({ status, signUp, history, location }) => (
+  <div className="w-11/12 md:w-9/12 lg:w-7/12 mx-auto m-16 rounded-lg flex items-stretch bg-red-400 h-100">
+    <div className="md:flex max-w-lg w-7/12 bg-color height items-center signup-card hidden rounded-l-lg ">
+      <div className=" pl-8 font-bold text-left text-white text-2xl">
+        <div>Welcome back,</div>
```
Comment (segunolalive) [inline src/pages/Signup/index.js]: Change `Welcome back,` to `Welcome to ErrorSwag`.

Wrap this text in an `h1` tag or the most appropriate heading tag. This is important for accessibility and SEO.

Use `<br>` tags to break the text into multiple lines if needed.

**Review by segunolalive — CHANGES_REQUESTED** [review]
🎊  This is good.

Make the correction on the SignUp form header and merge.
